{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The central dogma\n",
    "\n",
    "You are tasked with implementing in Python code a program that will translate a set of gene sequences (provided in CSV format) into protein sequences, and then search those protein sequences for a particular sequence of amino acids that represents a cleavage site for enterokinase. Enterokinase is a specific protease that cleaves after lysine (K) at its cleavage site DDDDK, however it will not will not cleave at a site followed by a proline (P).\n",
    "\n",
    "<table>\n",
    "<caption>Standard genetic code</caption>\n",
    "<tr>\n",
    "<th rowspan=\"2\">1st<br />\n",
    "base</th>\n",
    "<th colspan=\"8\">2nd base</th>\n",
    "<th rowspan=\"2\">3rd<br />\n",
    "base</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<th colspan=\"2\">T</th>\n",
    "<th colspan=\"2\">C</th>\n",
    "<th colspan=\"2\">A</th>\n",
    "<th colspan=\"2\">G</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<th rowspan=\"4\">T</th>\n",
    "<td>TTT</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#ffe75f\">(Phe/F) </td>\n",
    "<td>TCT</td>\n",
    "<td rowspan=\"4\" style=\"background-color:#b3dec0\">(Ser/S) </td>\n",
    "<td>TAT</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#b3dec0\">(Tyr/Y) </td>\n",
    "<td>TGT</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#b3dec0\">(Cys/C) </td>\n",
    "<th>T</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>TTC</td>\n",
    "<td>TCC</td>\n",
    "<td>TAC</td>\n",
    "<td>TGC</td>\n",
    "<th>C</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>TTA</td>\n",
    "<td rowspan=\"6\" style=\"background-color:#ffe75f\">(Leu/L) </td>\n",
    "<td>TCA</td>\n",
    "<td>TAA</td>\n",
    "<td style=\"background-color:#B0B0B0;\"> Stop</td>\n",
    "<td>TGA</td>\n",
    "<td style=\"background-color:#B0B0B0;\"> Stop</td>\n",
    "<th>A</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>TTG</td>\n",
    "<td>TCG</td>\n",
    "<td>TAG</td>\n",
    "<td style=\"background-color:#B0B0B0;\"> Stop</td>\n",
    "<td>TGG</td>\n",
    "<td style=\"background-color:#ffe75f;\">(Trp/W) &#160;&#160;&#160;&#160;</td>\n",
    "<th>G</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<th rowspan=\"4\">C</th>\n",
    "<td>CTT</td>\n",
    "<td>CCT</td>\n",
    "<td rowspan=\"4\" style=\"background-color:#ffe75f\">(Pro/P) </td>\n",
    "<td>CAT</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#bbbfe0\">(His/H) </td>\n",
    "<td>CGT</td>\n",
    "<td rowspan=\"4\" style=\"background-color:#bbbfe0\">(Arg/R) </td>\n",
    "<th>T</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>CTC</td>\n",
    "<td>CCC</td>\n",
    "<td>CAC</td>\n",
    "<td>CGC</td>\n",
    "<th>C</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>CTA</td>\n",
    "<td>CCA</td>\n",
    "<td>CAA</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#b3dec0\">(Gln/Q) </td>\n",
    "<td>CGA</td>\n",
    "<th>A</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>CTG</td>\n",
    "<td>CCG</td>\n",
    "<td>CAG</td>\n",
    "<td>CGG</td>\n",
    "<th>G</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<th rowspan=\"4\">A</th>\n",
    "<td>ATT</td>\n",
    "<td rowspan=\"3\" style=\"background-color:#ffe75f\">(Ile/I) </td>\n",
    "<td>ACT</td>\n",
    "<td rowspan=\"4\" style=\"background-color:#b3dec0\">(Thr/T) &#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;</td>\n",
    "<td>AAT</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#b3dec0\">(Asn/N) </td>\n",
    "<td>AGT</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#b3dec0\">(Ser/S) </td>\n",
    "<th>T</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>ATC</td>\n",
    "<td>ACC</td>\n",
    "<td>AAC</td>\n",
    "<td>AGC</td>\n",
    "<th>C</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>ATA</td>\n",
    "<td>ACA</td>\n",
    "<td>AAA</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#bbbfe0\">(Lys/K) </td>\n",
    "<td>AGA</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#bbbfe0\">(Arg/R) </td>\n",
    "<th>A</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>ATG<sup class=\"reference\" id=\"ref_methionineA\"></sup></td>\n",
    "<td style=\"background-color:#ffe75f;\">(Met/M) </td>\n",
    "<td>ACG</td>\n",
    "<td>AAG</td>\n",
    "<td>AGG</td>\n",
    "<th>G</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<th rowspan=\"4\">G</th>\n",
    "<td>GTT</td>\n",
    "<td rowspan=\"4\" style=\"background-color:#ffe75f\">(Val/V) </td>\n",
    "<td>GCT</td>\n",
    "<td rowspan=\"4\" style=\"background-color:#ffe75f\">(Ala/A) </td>\n",
    "<td>GAT</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#f8b7d3\">(Asp/D) </td>\n",
    "<td>GGT</td>\n",
    "<td rowspan=\"4\" style=\"background-color:#ffe75f\">(Gly/G) </td>\n",
    "<th>T</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>GTC</td>\n",
    "<td>GCC</td>\n",
    "<td>GAC</td>\n",
    "<td>GGC</td>\n",
    "<th>C</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>GTA</td>\n",
    "<td>GCA</td>\n",
    "<td>GAA</td>\n",
    "<td rowspan=\"2\" style=\"background-color:#f8b7d3\">(Glu/E) </td>\n",
    "<td>GGA</td>\n",
    "<th>A</th>\n",
    "</tr>\n",
    "<tr>\n",
    "<td>GTG</td>\n",
    "<td>GCG</td>\n",
    "<td>GAG</td>\n",
    "<td>GGG</td>\n",
    "<th>G</th>\n",
    "</tr>\n",
    "</table>\n",
    "\n",
    "\n",
    "## The task\n",
    "\n",
    "1) Write a Python 3 program that will translate each of the cDNA sequences into an amino acid sequence in single letter code. It should use the CSV format file \"cDNA\\_sequences.csv\" as input. The file is comma-delimited, and has the following layout (sequences are truncated in length for clarity):\n",
    "\n",
    "<table>\n",
    "<tr><td>ENST0000039045</td><td>ATGGTCCTGAAATTCTCC</td></tr>\n",
    "<tr><td>ENST0000055716</td><td>ATGGAGACTCTCCTGAAAGTGCTTT</td></tr>\n",
    "<tr><td>ENST0000039046</td><td>ATGGAGACTGTTCTGCAAGTACTCC</td></tr>\n",
    "</table>\n",
    "\n",
    "Amino acid (protein) sequences should be output as a **FASTA format file** named \"protein_sequences.fasta\". \n",
    "\n",
    "2) Your program should search the translated (amino acid) sequences for the enterokinase motif: DDDDK, but exclude those where the next amino acid after the motif is a proline - i.e. DDDDKP. How many of the sequences contain one or more enterokinase motifs?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Code Golf!\n",
    "\n",
    "Code Golf is a computer programming competition in which participants strive to achieve the shortest possible source code that implements an algorithm. \n",
    "\n",
    "The code that results from Code Golf naturally has some terrible stylistic problems such as lack of commenting, very short variable names and sometime obscure ways to solve problems. ** Please do not try to develop your real code in this style!**\n",
    "\n",
    "Code Golf is a very good exercise to make you think about other ways in which a task can be completed and understand the many alternative ways in which an algorithm can be written to solve a problem.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Use Code golf to restrict the number of characters used. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# timeit magic challenge \n",
    "\n",
    "You can use a variety of 'cell magics' to make use of additional features and functionality in Jupyter notebooks - see https://ipython.readthedocs.io/en/stable/interactive/magics.html for details. \n",
    "\n",
    "If you include %%timeit at the beginning of your cell, Jupyter will run the cell multiple times and tell you the mean taken to complete the program.  For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51.5 µs ± 25.6 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)\n"
     ]
    }
   ],
   "source": [
    "%%timeit\n",
    "# Find prime numbers\n",
    "def generate_primes(max_num):  \n",
    "    primes = []\n",
    "    for num in range(2, max_num + 1):\n",
    "        isPrime=True\n",
    "        for denominator in range(2, int(num ** 0.5) + 1):\n",
    "            if num % denominator == 0:\n",
    "                isPrime = False\n",
    "                break\n",
    "        if isPrime==True:\n",
    "            primes.append(num)\n",
    "    return(primes)\n",
    "generate_primes(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this exercise, instead of minimising the number of characters in your translation and sequence searching code, optimise it to run as fast as possible. Avoid looping through data structures repeatedly, use dictionaries instead of lists for look up and this about whether all the steps in your program are necessary and optimised. \n",
    "\n",
    "You can also repeat this type of optimisation for the answers for in Workshop 3 and 4 - here there is much more scope for optimising the code compared with the translation algorithm. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
